File: mssp/archaeology/017-cpython-zlib/src/island_test.py
"""The island test, run against the real zlib.

    python src/island_test.py

Section 3 is the control. Section 3b is the drill: a reader that DECLARES it
reports completeness while returning only bytes must be caught by running it.
"""
import json
import pathlib
import re
import sys
import zlib

from SMS import upstream
from TMS.readers import incremental, one_shot

HERE = pathlib.Path(__file__).parent
FMS = json.loads((HERE / "FMS" / "architecture.json").read_text(encoding="utf-8"))
POLICY = json.loads((HERE / "SCL" / "policy.json").read_text(encoding="utf-8"))
FAILURES = []
RAN = [0]


def check(label, ok, detail=""):
    RAN[0] += 1
    print(f'  {"PASS" if ok else "FAIL"}  {label}{f" - {detail}" if detail else ""}')
    if not ok:
        FAILURES.append(label)


print(f"\n  {upstream.versions()}")
whole, trunc = upstream.compressed(), upstream.truncated_stream()

print("\n== 1. each reader is an island, and FMS matches the tree")
for unit, declared in FMS["units"].items():
    directory = HERE.joinpath(*unit.split("/"))
    on_disk = sorted(p.name for p in directory.glob("*.py") if p.name != "__init__.py")
    check(f"{unit}: FMS declares what is on disk", on_disk == sorted(declared),
          f'disk {", ".join(on_disk)} | FMS {", ".join(sorted(declared))}')
    for name in on_disk:
        body = directory.joinpath(name).read_text(encoding="utf-8")
        check(f"{unit}/{name} imports nothing at all",
              not re.search(r"^\s*(import|from)\s", body, re.M))
check("both readers declare what they raise and what they report",
      all(isinstance(m.RAISES_ON_TRUNCATION, bool) and isinstance(m.REPORTS_COMPLETENESS, bool)
          for m in (one_shot, incremental)))

print("\n== 2. the same truncated stream, two readers from one module")
shot = upstream.one_shot(trunc)
step = upstream.incremental(trunc)
check("zlib.decompress raises", shot["raised"] is not None, shot["raised"])
check("  and the message names truncation", "truncated" in (shot["raised"] or ""))
check("  and it returns no bytes at all", shot["bytes"] is None)
check("decompressobj does NOT raise", step["raised"] is None)
check("  and returns real bytes", step["bytes"] is not None and len(step["bytes"]) > 0,
      f'{len(step["bytes"] or b"")} bytes')
check("  and reports eof = False", step["eof"] is False)
check("on a complete stream both succeed",
      upstream.one_shot(whole)["raised"] is None and upstream.incremental(whole)["raised"] is None)
check("and only then is eof True", upstream.incremental(whole)["eof"] is True)

print("\n== 3. the control - a complete stream carrying exactly those bytes")
control = upstream.incremental(upstream.control_for(step["bytes"]))
check("the control stream is complete", control["eof"] is True)
check("it decompresses to the same number of bytes", len(control["bytes"]) == len(step["bytes"]),
      f'{len(control["bytes"])} vs {len(step["bytes"])}')
check("BYTE-IDENTICAL to the truncated read", control["bytes"] == step["bytes"])
check("so no returned value can tell them apart", not (control["bytes"] != step["bytes"]))
check("and .eof can", control["eof"] != step["eof"], f'{control["eof"]} vs {step["eof"]}')
check("which means the discriminator is on the object, not on the value",
      incremental.COMPLETENESS_LIVES_ON.startswith("the decompressor object"))

print("\n== 3b. DRILL - a reader that overclaims must be caught by running it")


def reports_completeness(read):
    """Run the reader on a truncated stream and see whether the result carries
    a completeness answer at all."""
    return read(trunc)["eof"] is not None


liar_declares = True  # a unit claiming decompressobj's flag while returning only bytes
check("the drill unit declares it reports completeness", liar_declares is True)
check("running it says otherwise", reports_completeness(upstream.one_shot) is False)
check("so the declaration is refused", reports_completeness(upstream.one_shot) != liar_declares)
check("and the honest declarations hold under the same probe",
      reports_completeness(upstream.incremental) is incremental.REPORTS_COMPLETENESS
      and reports_completeness(upstream.one_shot) is one_shot.REPORTS_COMPLETENESS)
check("the raise-on-truncation declarations hold too",
      (upstream.one_shot(trunc)["raised"] is not None) is one_shot.RAISES_ON_TRUNCATION
      and (upstream.incremental(trunc)["raised"] is not None) is incremental.RAISES_ON_TRUNCATION)

print("\n== 4. a deliberate partial read lands in the same state as an accidental one")
capped = upstream.incremental(whole, max_length=100)
check("max_length returns exactly the cap", len(capped["bytes"]) == 100)
check("with eof False", capped["eof"] is False)
check("and an unconsumed tail", capped["unconsumed"] > 0, f'{capped["unconsumed"]} bytes')
check("the accidental truncation has eof False as well", step["eof"] is False)
check("so eof alone does not say WHY, only that it is not finished",
      capped["eof"] == step["eof"] and capped["unconsumed"] != (step["unconsumed"] or 0))

print("\n== 5. the truncation cuts mid-record")
complete_records, trailing = upstream.whole_records(step["bytes"])
check("the caller collects whole records", len(complete_records) > 0, f"{len(complete_records)}")
check("and one trailing fragment", len(trailing) > 0, repr(trailing))
check("the fragment starts like a record", trailing.startswith(b"record-"))
check("and is shorter than one", len(trailing) < len(b"record-000"))
check("a complete payload leaves no fragment",
      upstream.whole_records(upstream.PAYLOAD)[1] == b"")

print("\n== 6. what this does not change")
check("the incremental reader is right not to raise - not-finished-yet is its normal state",
      upstream.incremental(whole[: len(whole) // 4])["raised"] is None)
check("feeding it the rest still finishes",
      zlib.decompressobj().decompress(whole)[-11:] == b"record-059\n")
check("SCL names the cost in words",
      "one attribute read" in POLICY["what_this_costs"].lower())
check("and this deployment does refuse", POLICY["a_stream_that_did_not_reach_eof_is"] == "fatal")

print()
if FAILURES:
    print(f'  {len(FAILURES)} FAILED: {" | ".join(FAILURES)}')
    sys.exit(1)
print(f"  {RAN[0]} checks passed - every probe ran the real zlib")

File: mssp/examples/020-declare-capacity-not-state/src/island_test.mjs
// The island test.
//
//   node src/island_test.mjs
//
// Section 2 is the one that matters most: the challenge needs both arms,
// because a reader that answers the same way every time is right about one of
// them. Section 3 is the control pair — two silences that must not share a
// column.
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

import * as harness from "./SMS/challenge.mjs";
import * as policy from "./SCL/policy.mjs";
import * as claimsFraming from "./TMS/readers/claims_framing.mjs";
import * as framed from "./TMS/readers/framed.mjs";
import * as opaquePipe from "./TMS/readers/opaque_pipe.mjs";

const here = path.dirname(fileURLToPath(import.meta.url));
const CONTRACT = JSON.parse(fs.readFileSync(path.join(here, "FMS", "contract.json"), "utf8"));
const failures = [];
let ran = 0;
const check = (label, ok, detail = "") => {
  ran += 1;
  process.stdout.write(`  ${ok ? "PASS" : "FAIL"}  ${label}${detail ? ` - ${detail}` : ""}\n`);
  if (!ok) failures.push(label);
};
const say = (line = "") => process.stdout.write(`${line}\n`);

const { readers, problems } = harness.load();
const results = Object.fromEntries(
  Object.keys(readers).map((name) => [name, harness.challenge(readers[name])]));

say("\n== 1. every reader is an island, and FMS matches the tree");
check("loading raised no problems", problems.length === 0, problems.join("; "));
for (const [unit, declared] of Object.entries(CONTRACT.units)) {
  const dir = path.join(here, ...unit.split("/"));
  const onDisk = fs.readdirSync(dir).filter((n) => n.endsWith(".mjs")).sort();
  check(`${unit}: FMS declares what is on disk`,
    JSON.stringify(onDisk) === JSON.stringify([...declared].sort()),
    `disk ${onDisk.join(", ")} | FMS ${[...declared].sort().join(", ")}`);
  for (const file of onDisk) {
    const body = fs.readFileSync(path.join(dir, file), "utf8");
    check(`${unit}/${file} imports nothing at all`, !/^\s*import\s/m.test(body));
  }
}
check("every reader states a capacity either way",
  [framed, opaquePipe, claimsFraming].every((m) => typeof m.CLAIMS_CAN_DISCRIMINATE === "boolean"));
check("and says HOW, so the claim is inspectable as well as testable",
  [framed, opaquePipe, claimsFraming].every((m) => Boolean(m.HOW)));

say("\n== 2. the challenge needs both arms - a constant answer must not pass");
const alwaysTruncated = {
  NAME: "drill-always-truncated", CAN_FAIL_WITH: ["x"],
  CLAIMS_CAN_DISCRIMINATE: true, HOW: "it answers the same way every time",
  read: (stream) => ({ records: [{ from: "drill-always-truncated", id: "d-1" }],
    incomplete_because: "no terminator record" }),
};
const alwaysComplete = {
  NAME: "drill-always-complete", CAN_FAIL_WITH: ["x"],
  CLAIMS_CAN_DISCRIMINATE: true, HOW: "it answers the same way every time",
  read: (stream) => ({ records: [{ from: "drill-always-complete", id: "d-1" }],
    incomplete_because: null }),
};
const cheatTruncated = harness.challenge(alwaysTruncated);
const cheatComplete = harness.challenge(alwaysComplete);
check("a reader that always says truncated IS right about the truncated stream",
  cheatTruncated.rightAboutTruncated === true);
check("and wrong about the complete one", cheatTruncated.rightAboutComplete === false);
check("so a one-armed challenge would have passed it", cheatTruncated.constant === true);
check("and the two-armed one does not", cheatTruncated.passed === false);
check("the mirror image fails too", cheatComplete.passed === false
  && cheatComplete.rightAboutComplete === true && cheatComplete.rightAboutTruncated === false);
check("the harness knows both answers because it built both streams",
  harness.COMPLETE_STREAM.endsWith("<end>") && !harness.TRUNCATED_STREAM.endsWith("<end>"));
check("and the two streams differ only by the terminator",
  harness.COMPLETE_STREAM.replace("|<end>", "") === harness.TRUNCATED_STREAM);

say("\n== 3. the control pair - two silences that must not share a column");
const truncated = harness.readAll(readers, harness.TRUNCATED_STREAM);
const row = (name) => truncated.find((r) => r.reader === name);
const onComplete = harness.readAll(readers, harness.COMPLETE_STREAM);
check("on the truncated stream, framed speaks", row("framed").incomplete_because !== null);
check("and opaque-pipe does not", row("opaque-pipe").incomplete_because === null);
check("they return the same number of records",
  row("framed").records.length === row("opaque-pipe").records.length,
  `${row("framed").records.length}`);
const framedOnComplete = onComplete.find((r) => r.reader === "framed");
check("on the COMPLETE stream, framed is silent too - the control",
  framedOnComplete.incomplete_because === null);
check("so silence alone does not separate them",
  framedOnComplete.incomplete_because === row("opaque-pipe").incomplete_because);
check("and the completeness column does",
  framedOnComplete.completeness !== row("opaque-pipe").completeness,
  `${framedOnComplete.completeness} | ${row("opaque-pipe").completeness}`);
check("there are three values, and none of them is `complete`",
  new Set([harness.DECLARED_INCOMPLETE, harness.SILENT_AND_CAN_TELL, harness.SILENT_AND_CANNOT]).size === 3
  && ![harness.DECLARED_INCOMPLETE, harness.SILENT_AND_CAN_TELL, harness.SILENT_AND_CANNOT]
    .includes("complete"));

say("\n== 4. the liar is caught by running it, not by reading it");
check("claims-framing declares it can discriminate",
  claimsFraming.CLAIMS_CAN_DISCRIMINATE === true);
check("its source is identical in behaviour to the opaque pipe",
  claimsFraming.read(harness.TRUNCATED_STREAM).incomplete_because
  === opaquePipe.read(harness.TRUNCATED_STREAM).incomplete_because);
check("the challenge refuses it", results["claims-framing"].passed === false
  && harness.accepted(results["claims-framing"]) === false);
check("framed's identical claim is accepted", results.framed.passed === true
  && harness.accepted(results.framed) === true);
check("and opaque-pipe, which claimed nothing, is not refused for failing",
  harness.accepted(results["opaque-pipe"]) === true
  && results["opaque-pipe"].passed === false);
check("so refusal keys on the CLAIM, not on the outcome",
  results["claims-framing"].passed === results["opaque-pipe"].passed
  && harness.accepted(results["claims-framing"]) !== harness.accepted(results["opaque-pipe"]));

say("\n== 5. the floor now names what example 017's could not");
const kept = truncated.filter((entry) => harness.accepted(entry.challenge));
const numbers = harness.floor(kept);
check("the refused reader contributes nothing", kept.length === 2);
check("the floor counts what was declared", numbers.declared === 3, `${numbers.declared}`);
check("and separately counts what came from a blind reader", numbers.blind === 3, `${numbers.blind}`);
check("the two are different populations", numbers.declared + numbers.blind === numbers.total);
check("neither number is offered as a total of anything complete",
  !JSON.stringify(numbers).includes("complete"));

say("\n== 6. a capacity claim is not a state claim");
const vouching = {
  NAME: "drill-vouch", CAN_FAIL_WITH: ["x"], CLAIMS_CAN_DISCRIMINATE: true,
  HOW: "x", COMPLETE: true, read: () => ({ records: [], incomplete_because: null }),
};
check("DRILL: a reader declaring COMPLETE is still refused (改良點 15)",
  harness.load([vouching]).problems.some((p) => /declares COMPLETE/.test(p)));
const mute = {
  NAME: "drill-mute", CAN_FAIL_WITH: ["x"], HOW: "x",
  read: () => ({ records: [], incomplete_because: null }),
};
check("DRILL: a reader that claims no capacity either way is refused",
  harness.load([mute]).problems.some((p) => /does not claim a discrimination capacity/.test(p)));
check("but claiming FALSE is perfectly acceptable",
  harness.load().problems.length === 0 && opaquePipe.CLAIMS_CAN_DISCRIMINATE === false);
check("and no reader anywhere asserts it is complete",
  [framed, opaquePipe, claimsFraming].every((m) => m.COMPLETE === undefined));

say("\n== 7. what the challenge cannot do, asserted so it stays measured");
check("it only works where the harness can build a case it knows the answer to",
  policy.POLICY.what_this_cannot_reach.includes("challenge inputs the harness cannot construct"));
check("a reader that passes on THESE two streams has not been shown to pass on others",
  harness.COMPLETE_STREAM.split("|").length === 4,
  "one pair, one shape - the claim is demonstrated, not proved");
check("and a reader that disclaims capacity is trusted about its own blindness",
  results["opaque-pipe"].claimed === false && harness.accepted(results["opaque-pipe"]),
  "that direction is unchecked, and it is the conservative one - the same trade as 改良點 15");
check("the report never renders a blind reader's silence as informative",
  row("opaque-pipe").completeness === harness.SILENT_AND_CANNOT);

say("");
if (failures.length > 0) {
  say(`  ${failures.length} FAILED: ${failures.join(" | ")}`);
  process.exitCode = 1;
} else {
  say(`  ${ran} checks passed - ${Object.keys(readers).length} readers, 3 completeness values`);
}
